### PR TITLE
Use `virtual_modules` in Dune 1.7

### LINKED ISSUE #50: Doesn't work from utop : function `mirage_alloc_pages' is not available
```
$ docker run --rm -it ocaml/opam:debian-9_ocaml-4.04.0
$ opam install utop io-page-unix
$ utop
utop # #require "io-page-unix";;
Error: The external function `mirage_alloc_pages' is not available
utop # #require "io-page";;
Error: The external function `mirage_alloc_pages' is not available      
```

Is this supposed to work?

### PR DESCRIPTION
**Pull request by avsm**
This allows use of the interface file separately from the implementations. We now provide separate implementations that use dune variants so that they can be found via a variants tag search:

- `unix` (Unix or Windows)
- `js` (JavaScript)
- `xen` (Xen)

This cleans up the utop toplevel support as well to work by default now. (@avsm, fixes #50)
Also use implicit transitive dependencies, which will become the default in dune 2.0

I am noticing some local problems with opam installations though:
```
#=== ERROR while compiling io-page-unix.2.2.0 =================================#
# context     2.0.4 | macos/x86_64 | ocaml-base-compiler.4.07.1 | pinned(git+file:///Users/avsm/src/git/mirage/io-page#vmodules#a58f53d0)
# path        ~/.opam/ocaml-base-compiler.4.07.1/.opam-switch/build/io-page-unix.2.2.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p io-page-unix -j 27
# exit-code   1
# env-file    ~/.opam/log/io-page-unix-37329-03b1a4.env
# output-file ~/.opam/log/io-page-unix-37329-03b1a4.out
### output ###
# File "lib/unix/dune", line 7, characters 14-21:
# 7 |   (implements io-page)
#                   ^^^^^^^
# Error: Library implementation io-page for variant "unix" implements a library outside the project. This is forbidden.
```

which is why I am marking this as a draft PR to work through the CI first.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+
+  We now provide separate implementations that use dune variants
+  so that they can be found via a variants tag search:
+
+  - `unix` (Unix or Windows)
+  - `js` (JavaScript)
+  - `xen` (Xen)
+
+  This cleans up the utop toplevel support as well to work by
+  default now. (@avsm, fixes #50)
```
Comment (djs55) [inline CHANGES.md]: awesome!

**Comment by dinosaure** [comment]
Should related to mirage/irmin#338

**Comment by hannesm** [comment]
I still don't fully understand variants, but in the above list, what is happening with `freestanding`? or is this a non-issue? AFAIR the C stubs are provided by mirage-solo5 atm, but what does `dune` do when asked to use a variant that does _not_ exist for a certain package? take the "default" one (is there a default?)?

and what is the extensibility story - do all variants now need to end up in this `io-page` repository, or are they free to be defined elsewhere? (how will people/programs know about the existance elsewhere, what happens if there are multiple implementations of a single package for the same variant?)

thanks for working on this! and sorry for asking questions, but I could not discover where/whether they've been answered previously (please point me to a comment if this is the case)

**Comment by samoht** [comment]
Do we really need variants when we could just rely on dune doing the correct cross-compilation of C bindings?

**Comment by samoht** [comment]
Closing in favour of #60

**Pull request closed without merging**